### PR TITLE
Anomaly Detection Feature for WSO2 ML

### PR DESCRIPTION
**Pull request by ashensw**
This has been implemented based on spark k means algorithm.

### CODE REVIEW

File: modules/integration/tests-common/test-utils/src/main/java/org/wso2/carbon/ml/integration/common/utils/MLIntegrationTestConstants.java [hunk 1]
```diff
@@ -45,6 +45,13 @@
 	public static final String RESPONSE = "responseVariable";
 	public static final String ALGORITHM_NAME = "algorithmName";
 	public static final String ALGORITHM_TYPE = "algorithmType";
+	public static final String NORMAL_LABELS_CONFIG = "normalLabels";
+	public static final String NEW_NORMAL_LABEL_CONFIG = "newNormalLabel";
+	public static final String NEW_ANOMALY_LABEL_CONFIG = "newAnomalyLabel";
+	public static final String NORMALIZATION_CONFIG = "normalization";
+	public static final String RESPONSE_ATTRIBUTE_DEFAULT = "";
```
Comment (nirmal070125) [inline modules/integration/tests-common/test-utils/src/main/java/org/wso2/carbon/ml/integration/common/utils/MLIntegrationTestConstants.java]: Why empty strings as variables?

File: modules/integration/tests-common/test-utils/src/main/java/org/wso2/carbon/ml/integration/common/utils/MLIntegrationTestConstants.java [hunk 2]
```diff
@@ -45,6 +45,13 @@
...
+	public static final String TRAIN_DATA_FRACTION_DEFAULT = "1.0";
+	public static final String NORMAL_LABELS_DEFAULT = "";
```
Comment (nirmal070125) [inline modules/integration/tests-common/test-utils/src/main/java/org/wso2/carbon/ml/integration/common/utils/MLIntegrationTestConstants.java]: Why empty strings as variables?

File: modules/integration/tests-integration/tests-service/src/test/resources/testng.xml [hunk 3]
```diff
@@ -57,13 +57,13 @@
 		</packages>
 	</test>
 	<test name="ML Life-cycle-Tests" preserve-order="true" verbose="2"
-		parallel="false">
+		parallel="classes" thread-count="1">
```
Comment (nirmal070125) [inline modules/integration/tests-integration/tests-service/src/test/resources/testng.xml]: Do we need this change?

File: modules/samples/default/anomaly-detection-unlabeled-data/create-model-config [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+"value":"K_MEANS_ANOMALY_DETECTION_WITH_UNLABELED_DATA"
+},
+{
+"key":"algorithmType",
+"value":"Anomaly_Detection"
+},
+{
+"key":"responseVariable",
+"value":""
+},
```
Comment (nirmal070125) [inline modules/samples/default/anomaly-detection-unlabeled-data/create-model-config]: Do we need to send responseVariable here?

File: modules/samples/default/anomaly-detection-unlabeled-data/create-model-config [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+{
+"key":"trainDataFraction",
+"value":"1.0"
+},
+{
+"key":"normalLabels",
+"value":""
```
Comment (nirmal070125) [inline modules/samples/default/anomaly-detection-unlabeled-data/create-model-config]: why we send an empty normal label ?

**Comment by ashensw** [comment]
Modifications have been done. PR updated.

**Comment by nirmal070125** [comment]
Thanks @ashensw !!

File: modules/integration/tests-integration/tests-service/src/test/java/org/wso2/carbon/ml/MLTestUtils.java [hunk 6]
```diff
@@ -171,41 +175,111 @@ public static boolean checkModelStatus(String modelName, MLHttpClient mlHttpclie
...
+                setModelConfigurations(algorithmName, algorithmType, response, trainDataFraction));
+
+        // Set default Hyper-parameters
+        mlHttpclient.doHttpPost("/api/analyses/" + analysisId + "/hyperParams/defaults", null);
+
+        // Create a model
+        CloseableHttpResponse httpResponse = mlHttpclient.createModel(analysisId, versionSetId);
+        modelName = mlHttpclient.getModelName(httpResponse);
+        modelId = mlHttpclient.getModelId(modelName);
+
```
Comment (manoramahp) [inline modules/integration/tests-integration/tests-service/src/test/java/org/wso2/carbon/ml/MLTestUtils.java]: response.close();

Comment (ashensw) [inline modules/integration/tests-integration/tests-service/src/test/java/org/wso2/carbon/ml/MLTestUtils.java]: Thanks for pointing it out. Fixed with https://github.com/wso2/product-ml/pull/313/commits/12ea5260eb9f6401a97a560919fb4074235ebeaa

**Pull request merged**